In [1]:
import numpy as np
import csv
from sklearn.cross_validation import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.externals import joblib
from sklearn.metrics import roc_curve, auc
from sklearn.cross_validation import train_test_split
from sklearn.cross_validation import StratifiedKFold
from scipy import interp
from  matplotlib import pyplot as plt
from sklearn import cross_validation
import pandas as pd
from sklearn.metrics import roc_auc_score
from sklearn import metrics, preprocessing
import pickle
import matplotlib.pyplot as plt
import matplotlib.cm as cm
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import BernoulliNB

%matplotlib inline

In [2]:
##Read the data
df = pd.read_json("~/Documents/AML/HW2/train.json")

In [69]:
df2 = pd.read_json("~/Documents/AML/HW2/train.json")

In [71]:
cuisine_mapping = {0:"greek", 1:"southern_us", 2:"filipino", 3:"indian", 4:"jamaican", 5:"spanish", 6:"italian",\
                  7: "mexican", 8:"chinese", 9:"british", 10:"thai", 11:"vietnamese", 12:"cajun_creole",\
                  13:"brazilian", 14:"french", 15:"japanese", 16:"irish", 17:"korean", 
                   18:"moroccan", 19:"russian"}

In [70]:
df2.tail(5)

cuisine     id                                        ingredients
39769    irish  29109  [light brown sugar, granulated sugar, butter, ...
39770  italian  11462  [KRAFT Zesty Italian Dressing, purple onion, b...
39771    irish   2238  [eggs, citrus fruit, raisins, sourdough starte...
39772  chinese  41882  [boneless chicken skinless thigh, minced garli...
39773  mexican   2362  [green chile, jalapeno chilies, onions, ground...

In [4]:
df.head(5)

cuisine     id                                        ingredients
0        greek  10259  [romaine lettuce, black olives, grape tomatoes...
1  southern_us  25693  [plain flour, ground pepper, salt, tomatoes, g...
2     filipino  20130  [eggs, pepper, salt, mayonaise, cooking oil, g...
3       indian  22213                [water, vegetable oil, wheat, salt]
4       indian  13162  [black pepper, shallots, cornflour, cayenne pe...

### b)

In [5]:
## The number of recipes per cuisine genre are represented below by aggregating the recipes  per cuisine
numCuisineInst = df.groupby(['cuisine']).agg(['count'])
numCuisineInst 

id ingredients
             count       count
cuisine                       
brazilian      467         467
british        804         804
cajun_creole  1546        1546
chinese       2673        2673
filipino       755         755
french        2646        2646
greek         1175        1175
indian        3003        3003
irish          667         667
italian       7838        7838
jamaican       526         526
japanese      1423        1423
korean         830         830
mexican       6438        6438
moroccan       821         821
russian        489         489
southern_us   4320        4320
spanish        989         989
thai          1539        1539
vietnamese     825         825

In [6]:
## The number of categories is 20
len(numCuisineInst)

20

In [7]:
### Here we find that the number of rows is 39,774
ingredients_list = df['ingredients'].as_matrix()
len(ingredients_list)

39774

In [8]:
### Here we loop through every ingredient, and create a dictionary 
## of the ingredients to get only the unique ingredients
ingredientDict = dict()
for item in ingredients_list:
    for ingredient in item:
        if ingredient in ingredientDict:
            ingredientDict[ingredient] +=1
        else:
            ingredientDict[ingredient] =1
len(ingredientDict)

6714

The Dataset consists of 39,774 rows of data, with 20 different categories of cuisines. However, we find that there is quite a bit of redundancy in the data, with only 6714 differnet unique ingredients. If we examine the ingredients even further, we see that the ingredients suffer from casing and phrasing issues. For example, there might be 'Romaine' and 'romaine', and phrases like "Kraft Zesty Italian Dressing." So, go back and clean the ingredients so as to break apart any phrases into component words. 'Romaine Lettuce' becomes 'romaine', 'lettuce.' We also convert everything to lowercase, and drop any words that are 'a', 'the', 'of.' We then create a new dictionary from this and we find the number of unique ingredients to be about half as large at 3186

In [9]:
def clean_data(matrix):
    """ A function that accepts the matrix of ingredients as input. It loops through, splits any phrases into component
        parts, and adds these to a new array. Also added to this array is the ingredient converted to lower case. 
        Anything that has 'a', 'the', 'and', 'of' is dropped. Then, we add this array to a new matrix. This returns
        a matrix of lowercase, split, ingredients.
    """
    new_list = []
    for array in matrix:
        row = []
        for ingr in array:
            if ' ' in ingr:
#                 print(ingr)
                new_ingr_split = ingr.split()
#                 print(new_ingr_split)
                for item in new_ingr_split:
                    if item != 'the' and ingr != 'of' and ingr != 'and' and ingr != 'a':
                        row.append(item.lower())
            else:
                row.append(ingr.lower())
        new_list.append(row)
    return new_list

In [10]:
new_matrix = clean_data(ingredients_list)

In [11]:
## We Create a new dictionary from this cleaned data
ingr_Dict = dict()
for item in new_matrix:
    for ingredient in item:
        if ingredient in ingr_Dict:
            ingr_Dict[ingredient] +=1
        else:
            ingr_Dict[ingredient] =1

In [12]:
## the Number of unique ingredients in our cleaned data set
len(ingr_Dict)

3186

## C)

Next, we need to one hot encode the ingredients. To do this, we first convert all of the ingredients to numerical representations. Then we represent each of these numbers a binary vector.

In [13]:
## first, we create a dictionary of each item in the dataset, and map each ingredient as a key to a numerical value.
factorize_dict = dict()
k=0
for item in new_matrix:
    for ingredient in item:
        if ingredient not in factorize_dict:
            k+=1
            factorize_dict[ingredient] =k

In [14]:
#The lengths of the dictionary with frequency counts and numerical encoudings are the same
len(factorize_dict)

3186

In [15]:
## Now we go through the entire  matrix of ingredients, and replace each ingredient with the numerical value 
## In the dictionary we created above
factorized_list = []
for array in new_matrix:
    row = []
    for x in array:
        num = factorize_dict[x]
        row.append(num)
    factorized_list.append(row)

In [16]:
factorized_consolidated_list = list(factorize_dict.values())

In [17]:
## next we create an empty numpy matrix of zeroes. We will use this later to set specific index locations to 1
blank_data = np.empty((len(factorized_list), len(factorized_consolidated_list)+1))

In [18]:
##Now we loop through the entire factorized ingredient matrix represented with numerical values
## and set that index location in the corresponding row to 1
s = 0
while s < len(factorized_list):
    for item in factorized_list[s]:
        blank_data[s][item] = 1
    s+=1

In [19]:
##We also need to factorize the cuisines. Thankfully, pandas has a built in function
df['cuisine'] = pd.factorize(df['cuisine'])[0]

## D

In [20]:
##Now, we set our X = to the one encoded matrix of ingredients,
## and Y Values = to the cuisine labels represented as numbers
X= blank_data
Y = df.iloc[:,0].values

In [21]:
## We make sure the x and y shapes are the compatible
print ("X shape is: ", X.shape, "Y shape is" , Y.shape, )

X shape is:  (39774, 3187) Y shape is (39774,)


In [22]:
##make sure there are no nan values
np.any(np.isnan(X)), np.any(np.isnan(Y))

(False, False)

In [23]:
## Next we check that our one hot encoding worked, and that we have 1 values in our empty matrix
sum = 0
for item in X:
    for num in item: 
        sum = sum + num

In [24]:
print(sum)

752205.0


## E)

In [25]:
##Gaussian Naive Bayes
gnb = GaussianNB()
scores_gnb = cross_validation.cross_val_score(gnb, X, Y, cv=3)
scores_gnb.mean()

0.26502153877819484

In [27]:
##Bernouli Naive Bayes
bnb = BernoulliNB()
scores_bnb = cross_validation.cross_val_score(bnb, X, Y, cv=3)
scores_bnb.mean()

0.71317996873139122

If we look at naive Bayes, there is a drastic difference in performance. Bernoulli NB almost performs as well as Logistic Regression with a perfomance of .71. Gaussian on the other hand performs with a very poor rate of .26. A possible explanation is that the Bernoulli distribution more closely matches the Yummly data set, and Gaussian distriutions do not do a particularly good job of modeling the data. Additionally, Bernoulli Naive Bayes by its very nature is designed to only deal with binary values. Since our data is one hot encoded, it is enginerred more for Bernoulli then Gaussian. 

## F

In [28]:
scores = cross_validation.cross_val_score(LogisticRegression(), X, Y, cv=3)
scores.mean()

0.78199308839376502

We see as expected that with 3 Fold Cross Validation, Logistic Regression performs the best with projected perfomance of .78.

In [29]:
def evaluate(C):
    scores = cross_validation.cross_val_score(LogisticRegression(), X, Y, cv=3)
    return scores.mean()

In [30]:
def bestModel(allCs):
    """ Iterates through all of the C values we pass, and returns a tuple of the score of the best C,
    and the best C"""
    score = 0
    c_val = .001
    for c in allCs:
        currScore = evaluate(c)
        if currScore > score:
            score = currScore
            c_val = c
    return score, c_val

In [31]:
## We see that the best performing C is 1, which I have used above
possible_Cs = [0.001, 0.01, 0.1, 1, 10, 100, 1000, 10000]
bestC = bestModel(possible_Cs)
print(bestC)

(0.78199308839376502, 0.001)


## G)

In [32]:
clf_full = LogisticRegression()
clf_full.fit(X , Y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr',
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0)

In [33]:
## save model to load later
from sklearn.externals import joblib
joblib.dump(clf_full, 'log_reg_whats_cooking.pkl') 

['log_reg_whats_cooking.pkl',
 'log_reg_whats_cooking.pkl_01.npy',
 'log_reg_whats_cooking.pkl_02.npy',
 'log_reg_whats_cooking.pkl_03.npy']

In [36]:
## Load model
clf2_full = joblib.load('log_reg_whats_cooking.pkl') 
clf2_full.predict(X[50])

array([15])

## Cleaning the test data set

In [37]:
##Read the data
df_test = pd.read_json("~/Documents/AML/HW2/test.json")

In [38]:
df_test.head(5)

id                                        ingredients
0  18009  [baking powder, eggs, all-purpose flour, raisi...
1  28583  [sugar, egg yolks, corn starch, cream of tarta...
2  41580  [sausage links, fennel bulb, fronds, olive oil...
3  29752  [meat cuts, file powder, smoked sausage, okra,...
4  35687  [ground black pepper, salt, sausage casings, l...

In [39]:
ingredients_list_test = df_test['ingredients'].as_matrix()

In [40]:
new_matrix_test = clean_data(ingredients_list_test)

In [41]:
new_matrix_test[0]

['baking',
 'powder',
 'eggs',
 'all-purpose',
 'flour',
 'raisins',
 'milk',
 'white',
 'sugar']

In [42]:
print(len(new_matrix_test[1]))

20


In [45]:
factorized_list_test = []
numKeyErrors = 0
for array in new_matrix_test:
    row = []
    for x in array:
        if x in factorize_dict: 
            num = factorize_dict[x]
            row.append(num)
        else:
            numKeyErrors+=1
    factorized_list_test.append(row)

In [46]:
print(numKeyErrors)

145


In [47]:
factorized_list_test[0]

[72, 36, 22, 142, 18, 463, 27, 83, 64]

In [48]:
print(len(factorized_list_test[1]))

20


In [49]:
factorized_consolidated_list_test = list(factorize_dict.values())

In [50]:
## next we create an empty numpy matrix of zeroes. We will use this later to set specific index locations to 1
blank_data_test = np.empty((len(factorized_list_test), len(factorized_consolidated_list)+1))

In [51]:
print(blank_data_test[10:20])

[[ 0.  0.  0. ...,  0.  0.  0.]
 [ 0.  0.  0. ...,  0.  0.  0.]
 [ 0.  0.  0. ...,  0.  0.  0.]
 ..., 
 [ 0.  0.  0. ...,  0.  0.  0.]
 [ 0.  0.  0. ...,  0.  0.  0.]
 [ 0.  0.  0. ...,  0.  0.  0.]]


In [52]:
t = 0
while t < len(factorized_list_test):
    for item in factorized_list_test[t]:
        blank_data_test[t][item] = 1
    t+=1

In [53]:
X_test = blank_data_test

In [54]:
## We make sure the x and y shapes are the compatible
print ("X_test shape is: ", X_test.shape)

X_test shape is:  (9944, 3187)


In [55]:
np.any(np.isnan(X)), np.any(np.isnan(Y))

(False, False)

In [66]:
## Next we check that our one hot encoding worked, and that we have 1 values in our empty matrix
sum = 0
for item in X_test:
    for num in item: 
        sum = (sum + num)

In [67]:
print(sum)

188434.0


## Predicting the test data set

In [58]:
preds = clf2_full.predict(X_test)

In [72]:
preds_cuisines = [ cuisine_mapping[x] for x in preds]

In [73]:
print(preds_cuisines[0:2])

['british', 'southern_us']


In [74]:
left = pd.DataFrame(df_test['id'])

In [75]:
right = pd.DataFrame({"cuisine": preds_cuisines})

In [76]:
results = left.join(right)

In [77]:
results.head(5)

id       cuisine
0  18009       british
1  28583   southern_us
2  41580       italian
3  29752  cajun_creole
4  35687       italian

In [78]:
results.to_csv("predictions_whats_cooking_2.csv")